include/deal.II/base/index_set.h

```diff
@@ -79,6 +89,188 @@ class IndexSet
+     */
+    IntervalIterator(const IndexSet &idxset, size_type range_idx);
+
+    /**
+     * Construct an invalid iterator (used as end())
+     */
+    IntervalIterator(const IndexSet &idxset);
+
+    /**
+     * Prefix increment.
+     */
+    IntervalIterator &operator++ ();
+
+    /**
+     * Postfix increment.
+     */
+    IntervalIterator operator++ (int);
+
+    /**
+     * Dereferencing operator, returns an IntervalAccessor.
+     */
+    const IntervalAccessor &operator* () const;
+
+    /**
+     * Dereferencing operator, returns a pointer to an IntervalAccessor.
+     */
+    const IntervalAccessor *operator-> () const;
+
+    /**
+     * Comparison.
+     */
+    bool operator == (const IntervalIterator &) const;
+
+    /**
+     * Inverse of <tt>==</tt>.
+     */
+    bool operator != (const IntervalIterator &) const;
+
+    /**
+     * Comparison operator.
+     */
+    bool operator < (const IntervalIterator &) const;
+
+    /**
+     * Return the distance between the current iterator and the argument. The
+     * distance is given by how many times one has to apply operator++ to the
+     * current iterator to get the argument (for a positive return value), or
+     * operator-- (for a negative return value).
+     */
+    std::ptrdiff_t operator - (const IntervalIterator &p) const;
+
+  private:
+    IntervalAccessor accessor;
+  };
+
+  /**
+   * Class that represents an iterator pointing to a single element in the IndexSet as returned by IndexSet::begin().
+   */
+  class ElementIterator
+  {
+  public:
+    /**
+     * Construct an iterator pointing to the global index @p index in the interval @p range_idx
+     */
+    ElementIterator(const IndexSet &idxset, size_type range_idx, size_type index);
+    /**
+     * Construct an iterator pointing to the end of the IndexSet.
+     */
+    ElementIterator(const IndexSet &idxset);
+
+    /**
+     * Dereferencing operator. The returned value is the index of the element inside the IndexSet.
+     */
+    const size_type &operator* () const;
+
+    /**
+     * Does this iterator point to an existing element?
+     */
+    bool is_valid () const;
+
+    /**
+     * Prefix increment.
+     */
+    ElementIterator &operator++ ();
+
+    /**
+     * Postfix increment.
+     */
+    ElementIterator operator++ (int);
+
+
+    /**
+     * Comparison.
+     */
+    bool operator == (const ElementIterator &) const;
+
+    /**
+     * Inverse of <tt>==</tt>.
+     */
+    bool operator != (const ElementIterator &) const;
+
+    /**
+     * Comparison operator.
+     */
+    bool operator < (const ElementIterator &) const;
+
+    /**
+     * Return the distance between the current iterator and the argument. The
+     * distance is given by how many times one has to apply operator++ to the
+     * current iterator to get the argument (for a positive return value), or
+     * operator-- (for a negative return value).
```

Comment:
there is no operator--. I think the sentence is obvious, though, so do as you see fit.